platform/darwin/src/MGLConversion.h

```diff
@@ -0,0 +1,136 @@
+#import <Foundation/Foundation.h>
+
+#include <mbgl/util/logging.hpp>
+#include <mbgl/style/conversion.hpp>
+#include <mbgl/util/feature.hpp>
+#include <mbgl/util/optional.hpp>
+
+NS_ASSUME_NONNULL_BEGIN
+
+namespace mbgl {
+namespace style {
+namespace conversion {
+
+// A minimal wrapper class conforming to the requirements for `objectMember(v, name)` (see mbgl/style/conversion.hpp)
+// This is necessary because using `NSObject*` as the value type in `optional<NSObject*>` causes problems for the ARC,
+// due to things like `optional(const value_type& __v)`
+class OptionalNSObjectValue {
+public:
+    OptionalNSObjectValue(NSObject * _Nullable _value) : value(_value) {}
+    
+    explicit operator bool() const {
+        return value;
+    }
+    
+    NSObject * _Nullable operator*() {
+        NSCAssert(this, @"Expected non-null value.");
+        return value;
+    }
+private:
+    NSObject * _Nullable value;
+};
+
+inline bool isUndefined(const id value) {
+    return !value || value == [NSNull null];
+}
+
+inline bool isArray(const id value) {
+    return [value isKindOfClass:[NSArray class]];
+}
+
+inline bool isObject(const id value) {
+    return [value isKindOfClass:[NSDictionary class]];
+}
+
+inline std::size_t arrayLength(const id value) {
+    NSCAssert([value isKindOfClass:[NSArray class]], @"Value must be an NSArray for getLength().");
+    NSArray *array = value;
+    auto length = [array count];
+    NSCAssert(length < std::numeric_limits<size_t>::max(), @"Array length out of bounds.");
+    return length;
+}
+
+inline NSObject *arrayMember(const id value, std::size_t i) {
+    NSCAssert([value isKindOfClass:[NSArray class]], @"Value must be an NSArray for get(int).");
+    NSCAssert(i < NSUIntegerMax, @"Index must be less than NSUIntegerMax");
+    NSArray *array = value;
+    return array[i];
```

Comment:
Yes. It matters more for more complex functions where you’d otherwise potentially have to keep both variables in sync.